### PR TITLE
Small ventcrawling refactor

### PR DESCRIPTION
**Pull request by vincentiusvin**
## About The Pull Request
Most of our ventcrawling logic is a 5 year old port from /vg/station, i hoped to restructure it and make improvements. Shouldn't have any signficant behavior changes should it?

## Why It's Good For The Game
Made the ventcrawling code slightly simpler, makes room for future improvements (making it exit on alt-click and not be jank; cant seem to accomplish this myself).

## Changelog
:cl:
refactor: Refactored ventcrawling code. No expected behavior changes except for a few texts and the ability to crawl with passive vents.
/:cl:

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
Atomize out the hotkey stuff and I'll give you more gbp

File: code/__DEFINES/atmospherics.dm [hunk 1]
```diff
@@ -537,3 +537,8 @@ GLOBAL_LIST_INIT(pipe_paint_colors, sortList(list(
 #define EUPHORIA_INACTIVE 0
 #define EUPHORIA_ACTIVE 1
 #define EUPHORIA_LAST_FLAG 2
+
+// Ventcrawling bitflags
+#define VENTCRAWL_ALLOWED 					(1<<0)
+#define VENTCRAWL_ENTRANCE_ALLOWED 			(1<<1)
+#define VENTCRAWL_CAN_SEE	 				(1<<2)
```
Comment (Mothblocks) [inline code/__DEFINES/atmospherics.dm:544]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#tabs-not-spaces

File: code/modules/atmospherics/machinery/atmosmachinery.dm [hunk 2]
```diff
@@ -56,12 +56,15 @@
...
+	///Is the thing being rebuilt by SSair or not. Prevents list bloat
 	var/rebuilding = FALSE
 
+	//Default is to allow ventcrawling and seeing pipes
+	var/vent_movement = VENTCRAWL_ALLOWED | VENTCRAWL_CAN_SEE
+
 /obj/machinery/atmospherics/examine(mob/user)
 	. = ..()
-	if(is_type_in_list(src, GLOB.ventcrawl_machinery) && isliving(user))
+	if(src.vent_movement & VENTCRAWL_ENTRANCE_ALLOWED && isliving(user))
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/atmosmachinery.dm:67]:
```suggestion
	if((vent_movement & VENTCRAWL_ENTRANCE_ALLOWED) && isliving(user))
```

File: code/modules/atmospherics/machinery/atmosmachinery.dm [hunk 3]
```diff
@@ -439,40 +443,29 @@
...
+				playsound(src, 'sound/machines/ventcrawl.ogg', 50, TRUE, -3)
+
+	//Would be great if this could be implemented when someone alt-clicks the image.
+	if (target_move.vent_movement & VENTCRAWL_ENTRANCE_ALLOWED)
+		user.handle_ventcrawl(target_move)
 	//PLACEHOLDER COMMENT FOR ME TO READD THE 1 (?) DS DELAY THAT WAS IMPLEMENTED WITH A... TIMER?
 
 /obj/machinery/atmospherics/AltClick(mob/living/L)
-	if(istype(L) && is_type_in_list(src, GLOB.ventcrawl_machinery))
+	if(!src.vent_movement & VENTCRAWL_ALLOWED)
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/atmosmachinery.dm:462]:
```suggestion
	if(!(vent_movement & VENTCRAWL_ALLOWED))
```

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 4]
```diff
@@ -94,6 +94,7 @@
 	var/obj/item/radio/radio
 	var/radio_key = /obj/item/encryptionkey/headset_med
 	var/radio_channel = RADIO_CHANNEL_MEDICAL
+	vent_movement = 0
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:97]:
```suggestion
	vent_movement = NONE
```

File: code/modules/atmospherics/machinery/atmosmachinery.dm [hunk 5]
```diff
@@ -56,12 +56,15 @@
 	///Whether it can be painted
 	var/paintable = FALSE
 
-	///Is the thing being rebuilt by SSair or not. Prevents list blaot
+	///Is the thing being rebuilt by SSair or not. Prevents list bloat
 	var/rebuilding = FALSE
 
+	//Default is to allow ventcrawling and seeing pipes
+	var/vent_movement = VENTCRAWL_ALLOWED | VENTCRAWL_CAN_SEE
```
Comment (Mothblocks) [inline code/modules/atmospherics/machinery/atmosmachinery.dm:63]: Doc-comment this.

File: code/__DEFINES/atmospherics.dm (hunk 1 above)
Comment (Mothblocks) [reply code/__DEFINES/atmospherics.dm:544]: Also, document all these values.

File: code/modules/mob/living/ventcrawling.dm [hunk 6]
```diff
@@ -1,11 +1,11 @@
+// VENTCRAWLING
+// Handles the entrance and exit on ventcrawling
+/mob/living/proc/handle_ventcrawl(obj/machinery/atmospherics/components/A)
```
Comment (Mothblocks) [inline code/modules/mob/living/ventcrawling.dm:3]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/modules/atmospherics/machinery/atmosmachinery.dm (hunk 3 above)
Comment (Mothblocks) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:462]: This didn't work, by the way. `!vent_movement` is calculated first.

image

File: code/modules/mob/living/ventcrawling.dm [hunk 7]
```diff
@@ -25,97 +25,68 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
+			visible_message("<span class='notice'>[src] scrambles out from the ventilation ducts!</span>","<span class='notice'>You out from the ventilation ducts.</span>")
+			forceMove(A.loc)
+			REMOVE_TRAIT(src, TRAIT_MOVE_VENTCRAWLING, VENTCRAWLING_TRAIT)
+			update_pipe_vision()
+		//Entrance here
+		else
+			var/datum/pipeline/vent_parent = A.parents[1]
+			if(vent_parent && (vent_parent.members.len || vent_parent.other_atmosmch))
+				visible_message("<span class='notice'>[src] begins climbing into the ventilation system...</span>" ,"<span class='notice'>You begin climbing into the ventilation system...</span>")
+				if(!do_after(src, 25, target = A))
```
Comment (Mothblocks) [inline code/modules/mob/living/ventcrawling.dm:51]: `2.5 SECONDS`.

File: code/modules/mob/living/ventcrawling.dm [hunk 8]
```diff
@@ -25,97 +25,68 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
-	return
-
-/mob/living/update_pipe_vision(atom/new_loc = null)
-	. = loc
-	if(new_loc)
-		. = new_loc
-	remove_ventcrawl()
-	add_ventcrawl(.)
+/mob/living/proc/update_pipe_vision()
+	if(HAS_TRAIT(src, TRAIT_MOVE_VENTCRAWLING) && istype(src.loc, /obj/machinery/atmospherics) && movement_type & VENTCRAWLING)
```
Comment (Mothblocks) [inline code/modules/mob/living/ventcrawling.dm:70]:
```suggestion
	if(HAS_TRAIT(src, TRAIT_MOVE_VENTCRAWLING) && istype(src, /obj/machinery/atmospherics) && movement_type & VENTCRAWLING)
```

File: code/modules/mob/living/ventcrawling.dm [hunk 9]
```diff
@@ -25,97 +25,68 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
+		var/list/totalMembers = list()
+		var/obj/machinery/atmospherics/current_location = src.loc
+		for(var/datum/pipeline/P in current_location.returnPipenets())
```
Comment (Mothblocks) [inline code/modules/mob/living/ventcrawling.dm:73]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/modules/mob/living/ventcrawling.dm [hunk 10]
```diff
@@ -25,97 +25,68 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
-/mob/living/update_pipe_vision(atom/new_loc = null)
-	. = loc
-	if(new_loc)
-		. = new_loc
-	remove_ventcrawl()
-	add_ventcrawl(.)
+/mob/living/proc/update_pipe_vision()
+	if(HAS_TRAIT(src, TRAIT_MOVE_VENTCRAWLING) && istype(src.loc, /obj/machinery/atmospherics) && movement_type & VENTCRAWLING)
+		var/list/totalMembers = list()
+		var/obj/machinery/atmospherics/current_location = src.loc
```
Comment (Mothblocks) [inline code/modules/mob/living/ventcrawling.dm:72]:
```suggestion
		var/obj/machinery/atmospherics/current_location = loc
```

File: code/modules/mob/living/ventcrawling.dm [hunk 11]
```diff
@@ -25,97 +25,68 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
-
-/mob/living/update_pipe_vision(atom/new_loc = null)
-	. = loc
-	if(new_loc)
-		. = new_loc
-	remove_ventcrawl()
-	add_ventcrawl(.)
+/mob/living/proc/update_pipe_vision()
+	if(HAS_TRAIT(src, TRAIT_MOVE_VENTCRAWLING) && istype(src.loc, /obj/machinery/atmospherics) && movement_type & VENTCRAWLING)
+		var/list/totalMembers = list()
```
Comment (Mothblocks) [inline code/modules/mob/living/ventcrawling.dm:71]: Rename it `total_members`.

File: code/modules/mob/living/ventcrawling.dm [hunk 12]
```diff
@@ -25,97 +25,68 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
-/atom/proc/update_pipe_vision(atom/new_loc = null)
-	return
-
-/mob/living/update_pipe_vision(atom/new_loc = null)
-	. = loc
-	if(new_loc)
-		. = new_loc
-	remove_ventcrawl()
-	add_ventcrawl(.)
+/mob/living/proc/update_pipe_vision()
```
Comment (Mothblocks) [inline code/modules/mob/living/ventcrawling.dm:69]: Document this.

File: code/modules/mob/living/ventcrawling.dm [hunk 13]
```diff
@@ -25,97 +25,68 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
+		var/obj/machinery/atmospherics/current_location = src.loc
+		for(var/datum/pipeline/P in current_location.returnPipenets())
+			totalMembers += P.members
+			totalMembers += P.other_atmosmch
+
+		if(!totalMembers.len)
+			return
+
+		if(client)
+			for(var/obj/machinery/atmospherics/A in totalMembers)
```
Comment (Mothblocks) [inline code/modules/mob/living/ventcrawling.dm:81]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/modules/mob/living/ventcrawling.dm [hunk 14]
```diff
@@ -25,97 +25,68 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
+				if(A.vent_movement & VENTCRAWL_CAN_SEE && in_view_range(client.mob, A))
```
Comment (Mothblocks) [inline code/modules/mob/living/ventcrawling.dm:82]: Early-continue.

File: code/modules/mob/living/ventcrawling.dm [hunk 15]
```diff
@@ -25,97 +25,68 @@ GLOBAL_LIST_INIT(ventcrawl_machinery, typecacheof(list(
...
+					if(!A.pipe_vision_img)
+						A.pipe_vision_img = image(A, A.loc, layer = ABOVE_HUD_LAYER, dir = A.dir)
+						A.pipe_vision_img.plane = ABOVE_HUD_PLANE
+					client.images += A.pipe_vision_img
+					pipes_shown += A.pipe_vision_img
+	else
+		if(client)
```
Comment (Mothblocks) [inline code/modules/mob/living/ventcrawling.dm:89]: Merge these into one `else if (!isnull(client))`.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by Mothblocks** [comment]
There's some more `src.` I didn't put in there by the way, `src` is implicit.

File: code/modules/atmospherics/machinery/atmosmachinery.dm (hunk 5 above)
Comment (vincentiusvin) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:63]: Did all the other suggested changes but I'm not sure what you meant by documenting this. Let me know if the comment on top is not enough / you want me to add to it.

File: code/modules/mob/living/ventcrawling.dm (hunk 8 above)
Comment (vincentiusvin) [reply code/modules/mob/living/ventcrawling.dm]: It was meant to check for the location var of the mob.

**Comment by vincentiusvin** [comment]
Alright, thanks for the review! 
Tested locally, the behavior works fine again. Let me know if there are more problems.

Sorry i didn't commit your suggestions, had to go through one by one to keep my mind straight.

**Comment by vincentiusvin** [comment]
Hang on, on second thought i have a suspicion that this doesn't delete the displayed images of another pipenet before a mob leaves. I will address this tomorrow.

File: code/modules/atmospherics/machinery/atmosmachinery.dm (hunk 5 above)
Comment (Mothblocks) [reply code/modules/atmospherics/machinery/atmosmachinery.dm:63]: Doc comments are `///`, but this doesn't describe what this is, just what the defaults are.

File: code/__DEFINES/atmospherics.dm [hunk 16]
```diff
@@ -537,3 +537,8 @@ GLOBAL_LIST_INIT(pipe_paint_colors, sortList(list(
 #define EUPHORIA_INACTIVE 0
 #define EUPHORIA_ACTIVE 1
 #define EUPHORIA_LAST_FLAG 2
+
+// Ventcrawling bitflags, handled in var/vent_movement
+#define VENTCRAWL_ALLOWED	(1<<0) //Allows for ventcrawling to occur. All atmospheric machines have this flag on by default. Cryo is the exception
+#define VENTCRAWL_ENTRANCE_ALLOWED	(1<<1) //Allows mobs to enter or leave from atmospheric machines. On for passive, unary, and scrubber vents.
+#define VENTCRAWL_CAN_SEE	(1<<2) //Used to check if a machinery is visible. Called by update_pipe_vision(). On by default for all except cryo.
```
Comment (Mothblocks) [inline code/__DEFINES/atmospherics.dm:544]: These aren't doc comments.

Doc comments precede defines with `///`.

File: code/__DEFINES/atmospherics.dm [hunk 17]
```diff
@@ -539,6 +539,9 @@ GLOBAL_LIST_INIT(pipe_paint_colors, sortList(list(
 #define EUPHORIA_LAST_FLAG 2
 
 // Ventcrawling bitflags, handled in var/vent_movement
-#define VENTCRAWL_ALLOWED	(1<<0) //Allows for ventcrawling to occur. All atmospheric machines have this flag on by default. Cryo is the exception
-#define VENTCRAWL_ENTRANCE_ALLOWED	(1<<1) //Allows mobs to enter or leave from atmospheric machines. On for passive, unary, and scrubber vents.
-#define VENTCRAWL_CAN_SEE	(1<<2) //Used to check if a machinery is visible. Called by update_pipe_vision(). On by default for all except cryo.
+///Allows for ventcrawling to occur. All atmospheric machines have this flag on by default. Cryo is the exception
+#define VENTCRAWL_ALLOWED	(1<<0)
+///Allows mobs to enter or leave from atmospheric machines. On for passive, unary, and scrubber vents.
+#define VENTCRAWL_ENTRANCE_ALLOWED	(1<<1)
```
Comment (Mothblocks) [inline code/__DEFINES/atmospherics.dm:545]:
```suggestion
#define VENTCRAWL_ENTRANCE_ALLOWED (1<<1)
```

**Pull request merged**